### DISCUSSION TITLE
Final touches to the factory API for 0.5

### DISCUSSION THREAD

**Discussion by AaronErhardt** [Relm4/Relm4#191] [Polls]
The factory API was subject to major changes for 0.5. Each element is now an full-featured component. However, there are still some limitations and problems: 

1. By using `RefCell` internally, we are not able to implement `Index` or `IndexMut` (the [] operator), which means we need `get()` and `get_mut()`.
2. If you forget to call `render_changes()`, the factory does nothing. The most straight-forward approach to prevent this would be using a type similar to `MutexGuard` that allows us to "lock" the factory and perform changes that are rendered automatically as soon as the type is dropped.
3. `render_changes()` only takes `&self` to make it possible to call it from `update_view()`. This adds quite a bit of internal complexity and could be fixed by the guard-type (see 2.) too.

To fix 1. we will need a bit of `unsafe` code. Yet I think we can argue that it is safe to say that at any point in the program the data of a factory component is either owned by itself or the parent. At no point it is accessed by both, which should allow us to give full ownership to the update functions of the component and its parent without violating the ownership rules. This will also make 2. easier to fix.

If anyone has other suggestions, let us know.

**Comment by songww** [comment]
I'm not biased against unsafe, and the gtk we rely on is built on top of a lot of unsafe. But it is true that there is a lot of dislike for unsafe.

My English is not good, this is from deepl translate

**Comment by AaronErhardt** [comment]
The third solution won the poll and turned out pretty well: https://github.com/Relm4/Relm4/pull/192
